### DISCUSSION TITLE
Calculating grain boundaries on big EBSD datasets faster

### DISCUSSION THREAD

**Discussion by meganhurley13** [mtex-toolbox/mtex#2029] [Ask Anything]
Hello!

I am wondering if anyone has experience or workarounds for using MTEX to work with big datasets, particularly when it comes to speed and memory efficiency when calculating grain boundaries.

For my research, I have been using the parent grain reconstruction code to identify parent austenite grains and annealing twin from collected martensite EBSD. In order to get enough austenite grains while maintaining a high enough resolution for martensite, I have to collect EBSD over a large area with a fairly small step size, causing the datasets to be several gigabytes and contain upwards of ~30 million EBSD points. Due to this size, I am unable to run the full data set through either the initial calcGrains or the variant voting sections without causing MATLAB to run out of memory or simply crash. For more context, this error would always occur on my Microsoft Surface w/ 16GB RAM, but would also happen on one of my research group's remote workstations, which has ~192GB of virtual RAM. On this computer, I have been able to work around this issue by breaking the dataset into smaller sections. While this method works to allow for the parent EBSD variables to be concatenated together, it does not work the same for the identified parent grains. Performing calcGrains again on the entire reconstructed EBSD region does tend to work fine, however it takes a very long time (taking nearly 2 full weeks in some instances!).

So if there was a way that I could calcGrains on the reconstructed EBSD in sections, and then concatenated all calculated grains into one variable that would be ideal. But I am not sure if that is possible.

If there are any suggestions or recommendations, please let me know!

Thanks,
Megan

**Comment by AzdiarGazder** [comment]
Hi @meganhurley13,

Since memory restrictions are causing your large maps to crash, a workaround may be the only expedient way forward.
Perhaps the following workflow may help you succeed:
1. Use the [split](https://github.com/AzdiarGazder/mtexTools/tree/main/split) function to sub-divide your large map into a reasonable number of sub-maps.
2. Apply [parent grain reconstruction](https://github.com/ORTools4MTEX/ORTools) on each sub-map.
3. Use the [stitch](https://github.com/AzdiarGazder/mtexTools/tree/main/stitch) function to merge the reconstructed sub-maps back into one large map.

Hope this helps.

Warm regards,
Azdi

**Comment by psmgamerhell** [comment]
Hi Megan and Azdi,

I think it could be interesting to check with developper if it's relevant to begin to adapt Data Type of all variables used in the source code.
Typically using double float precision by default use twice as much memory than single float precision and same for integer.

By using the command whos in matlab you can visuallise memory usage by variable and type.
And we can see that almost all variables used in MTEX are in double precision consuming a lot memory.

I'm starting to adapt the code from the MTEX v5.11.2
Warms regards,
François
